### ISSUE TITLE
Race condition in acceptance tests for Ember.js 2.15

### ISSUE THREAD

**Issue by YoranBrondsema** [sreedhar7/ember-app-scheduler#11]
I was encountering the same issue as https://github.com/dollarshaveclub/ember-router-scroll/issues/72. I think the error lies in this addon rather than `ember-router-scroll`.

1. The origin is https://github.com/dollarshaveclub/ember-router-scroll/blob/2d577cd4a695310e02db1217d9f8eff5f9bdf1d1/addon/index.js#L25, which explains why it happens only in acceptance tests as route transitions normally do not occur during unit and component integration tests.
2. The error happens at random times during the execution of acceptance tests which seems to suggest that it's a race condition (the most fun ones :/).

Armed with a debugger and enough patience for the bug to happen, I now have a hypothesis for what is going on. The logic inside the scheduler service is quite complex so I might be wrong. I'll need to test with a fix on my test suite to get confidence that it indeed is the problem.

1. The error happens on https://github.com/sreedhar7/ember-app-scheduler/blob/master/addon/services/scheduler.js#L71. `this.queues` is `null` so the access `this.queues[queueName]` is the thing that is throwing the error. From the state I could see that the `isDestroyed` property on the service was `true`. This means that the `willDestroy` callback has run, which indeed sets `this.queues` to `null`.
2. The call to `flushQueue` comes from https://github.com/sreedhar7/ember-app-scheduler/blob/master/addon/services/scheduler.js#L145.

Now is the question, how come the line at https://github.com/sreedhar7/ember-app-scheduler/blob/master/addon/services/scheduler.js#L145 was executed in the first place? Since the service was destroyed, the `rAF` callback should have been cancelled (from https://github.com/sreedhar7/ember-app-scheduler/blob/master/addon/services/scheduler.js#L163).

I think the only possible explanation is that the following order of execution happens:
1. The last `requestAnimationFrame` callback is executed (on https://github.com/sreedhar7/ember-app-scheduler/blob/master/addon/services/scheduler.js#L114).
2. The service is destroyed.
3. https://github.com/sreedhar7/ember-app-scheduler/blob/master/addon/services/scheduler.js#L145 is executed via https://github.com/sreedhar7/ember-app-scheduler/blob/master/addon/services/scheduler.js#L128.

The `rAF` callback is executed in a `run.later`. I'm thinking that between those two moments, the service is destroyed. Therefore, `this.queues` is `null` by the time the callback is called.

It's kind of hard to explain but my fix proposal is to add a guard as follows:

````javascript
_rAFCallback(resolve) {
  this._nextPaintTimeout = run.later(() => {
    if (!this.isDestroyed) {
      this._nextAfterPaintPromise = null;
      this._nextPaintFrame = null;
      this._nextPaintTimeout = null;
      resolve();
    }
  }, 0)
},
````

The reason being, I think that the service is destroyed between the scheduling of the callback through `run.later` and its execution.

Like I said, I still have to test on a fork to see if my explanation is correct and actually fixes the issue. But I thought I'd throw it out here to see if someone had any comments or ideas.

Also, I have no idea why this happens only on Ember.js 2.15.

**Comment by YoranBrondsema** [comment]
Sorry, I got it all wrong... I realized later that `ember-router-scroll` uses version 0.0.5 of this addon and the changes in https://github.com/sreedhar7/ember-app-scheduler/commit/e7fd0749610af72873ed4ab8ea1ad060321d040c seem to fix it.

https://github.com/dollarshaveclub/ember-router-scroll/pull/73 does the upgrade to 0.2.0 in `ember-router-scroll`.

My apologies for any time wasted looking at this issue.
